File: advent_of_code/2021/day20/main.cpp
#include <bits/stdc++.h>
using namespace std;

string dic = "";
vector<string> img(0, "");
bool flipped = false;

char conv(int x, int y) {
	int idx = 0;
	int cnt = 9;
	for (int i = x-1; i <= x+1; i++)
		for (int j = y-1; j <= y+1; j++) {
			cnt--;
			if (0 <= i && i < (int)img.size() && 0 <= j && j < (int)img[0].size()) {
				if (img[i][j] == '#')
					idx |= 1<<cnt;
			} else if (flipped)
				idx |= 1<<cnt;
		}
	return dic[idx];
}

void printImg(vector<string>& s) {
	for (string& a : s)
		cout << a << endl;
	cout << endl;
}

int count(vector<string>& s) {
	int res = 0;
	for (string& a : s)
		for (char c : a)
			res += c == '#';
	return res;
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(NULL);
	cin >> dic;
	string s;
	while (cin >> s)
		img.push_back(s);
	for (int k = 0; k < 50; k++) {
		vector<string> img2(0, "");
		for (int i = -1; i <= (int)img.size(); i++) {
			img2.push_back("");
			for (int j = -1; j <= (int)img[0].size(); j++)
				img2.back().push_back(conv(i, j));
		}
		if (!flipped && dic[0] == '#')
			flipped = true;
		else if (flipped && dic[511] == '.')
			flipped = false;
		img = img2;
	}
	printf("%d\n", count(img));
}

File: advent_of_code/2021/day20/main.py
dic = input()
input()
img = []
while True:
	s = input()
	if s == "":
		break
	img.append(s)

def printImg(img):
	for s in img:
		print(s)
	print()

for k in range(2):
	img2 = []
	for i in range(-1, len(img)+1):
		img2.append("")
		for j in range(-1, len(img[0])+1):
			cnt = -1
			idx = 0
			for x in range(i+1, i-2, -1):
				for y in range(j+1, j-2, -1):
					cnt += 1
					print("cnt:", cnt)
					if 0 <= x and x < len(img) and 0 <= y and y < len(img[0]):
						if img[x][y] == '#':
							idx |= 1<<cnt
			print(idx)
			img2[-1] += dic[idx]
	printImg(img)
	img = img2
printImg(img)

res = 0
for s in img:
	for c in s:
		res += c == '#'
print(res)

